### PR TITLE
Colless

### PR DESCRIPTION
**Pull request by bcngod**

### CODE REVIEW

File: contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    spawned(asset, msg.sender)
+    nonReentrant
+    {
+        // check usefulness of tx
+        require(mainAmount != 0, "Unit Protocol: USELESS_TX");
+
+        uint debt = vault.debts(asset, msg.sender);
+        require(debt != 0 && usdpAmount != debt, "Unit Protocol: USE_REPAY_ALL_INSTEAD");
+
+        if (mainAmount != 0) {
```
Comment (Eenae) [inline contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol:174]: The check can be omitted

File: contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    spawned(vault.weth(), msg.sender)
+    nonReentrant
+    {
+        // check usefulness of tx
+        require(ethAmount != 0, "Unit Protocol: USELESS_TX");
+
+        uint debt = vault.debts(vault.weth(), msg.sender);
+        require(debt != 0 && usdpAmount != debt, "Unit Protocol: USE_REPAY_ALL_INSTEAD");
+
+        if (ethAmount != 0) {
```
Comment (Eenae) [inline contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol:213]: The check can be omitted

File: contracts/vault-managers/VaultManagerKeep3rPoolTokenBase.sol [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    spawned(asset, msg.sender)
+    nonReentrant
+    {
+        // check usefulness of tx
+        require(mainAmount != 0, "Unit Protocol: USELESS_TX");
+
+        uint debt = vault.debts(asset, msg.sender);
+        require(debt != 0 && usdpAmount != debt, "Unit Protocol: USE_REPAY_ALL_INSTEAD");
+
+        if (mainAmount != 0) {
```
Comment (Eenae) [inline contracts/vault-managers/VaultManagerKeep3rPoolTokenBase.sol:136]: The check can be omitted

**Review by Eenae — COMMENTED** [review]
Only minor comments.

File: contracts/vault-managers/VaultManagerKeep3rPoolTokenBase.sol [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        require(vault.oracleType(asset, user) == ORACLE_TYPE, "Unit Protocol: WRONG_ORACLE_TYPE");
+        _;
+    }
+
+    /**
+     * @param _vaultManagerParameters The address of the contract with vault manager parameters
+     * @param _keep3rPoolToken The address of Keep3r-based Oracle for pool tokens
+     * @param _oracleType The oracle type ID
+     **/
+    constructor(address payable _vaultManagerParameters, address _keep3rPoolToken, uint _oracleType) public {
```
[the remark this answers is not in the document]
Comment (bcngod) [reply contracts/vault-managers/VaultManagerKeep3rPoolTokenBase.sol]: No, we don't. Removed.

File: contracts/vault-managers/VaultManagerKeep3rUniswapPoolToken.sol [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * @title VaultManagerKeep3rUniswapPoolToken
+ **/
+contract VaultManagerKeep3rUniswapPoolToken is VaultManagerKeep3rPoolTokenBase {
+
+    /**
+     * @param _vaultManagerParameters The address of the contract with vault manager parameters
+     * @param _keep3rPoolToken The address of Keep3r-based Oracle for pool tokens
+     **/
+    constructor(address payable _vaultManagerParameters, address _keep3rPoolToken)
```
[the remark this answers is not in the document]
Comment (bcngod) [reply contracts/vault-managers/VaultManagerKeep3rUniswapPoolToken.sol]: Removed

File: contracts/vault-managers/VaultManagerStandard.sol [hunk 6]
```diff
@@ -38,51 +37,40 @@ contract VaultManagerStandard is ReentrancyGuard {
...
-     * @param colAmount The amount of COL token to deposit
      **/
-    function deposit(address asset, uint mainAmount, uint colAmount) public nonReentrant {
+    function deposit(address asset, uint mainAmount) public nonReentrant {
 
         // check usefulness of tx
-        require(mainAmount != 0 || colAmount != 0, "Unit Protocol: USELESS_TX");
+        require(mainAmount != 0, "Unit Protocol: USELESS_TX");
 
         if (mainAmount != 0) {
```
[the remark this answers is not in the document]
Comment (bcngod) [reply contracts/vault-managers/VaultManagerStandard.sol]: Omitted

File: contracts/vault-managers/VaultManagerStandard.sol [hunk 7]
```diff
@@ -38,51 +37,40 @@ contract VaultManagerStandard is ReentrancyGuard {
...
-     * @param colAmount The amount of COL token to deposit
      **/
-    function deposit_Eth(uint colAmount) public payable nonReentrant {
+    function deposit_Eth() public payable nonReentrant {
 
         // check usefulness of tx
-        require(msg.value != 0 || colAmount != 0, "Unit Protocol: USELESS_TX");
+        require(msg.value != 0, "Unit Protocol: USELESS_TX");
 
         if (msg.value != 0) {
```
[the remark this answers is not in the document]
Comment (bcngod) [reply contracts/vault-managers/VaultManagerStandard.sol]: Omitted

File: contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol (hunk 1 above)
Comment (bcngod) [reply contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol]: Omitted

File: contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol (hunk 2 above)
Comment (bcngod) [reply contracts/vault-managers/VaultManagerKeep3rMainAssetBase.sol]: Omitted

File: contracts/vault-managers/VaultManagerKeep3rPoolTokenBase.sol (hunk 3 above)
Comment (bcngod) [reply contracts/vault-managers/VaultManagerKeep3rPoolTokenBase.sol]: Omitted

**Review by Eenae — APPROVED** [review]

**Pull request merged**